Sources/Rendering/OpenGL/Texture/index.js

```diff
@@ -483,10 +483,10 @@ function vtkOpenGLTexture(publicAPI, model) {
         //   return model.context.BYTE;
         case VtkDataTypes.UNSIGNED_CHAR:
           return model.context.UNSIGNED_BYTE;
-        // case VtkDataTypes.SHORT:
-        //   return model.context.SHORT;
-        // case VtkDataTypes.UNSIGNED_SHORT:
-        //   return model.context.UNSIGNED_SHORT;
+        case VtkDataTypes.SHORT:
+          return model.context.HALF_FLOAT;
+        case VtkDataTypes.UNSIGNED_SHORT:
+          return model.context.HALF_FLOAT;
```

Comment:
OK I spoke too soon. We need to think about this section right here. The issue is that float can fully represent short and unsigned short while half float cannot. 

Short has 15 bits of precision versus about 11 in half float. Unsigned short has 16 bits versus half floats 11. So while we can stuff short and short and unsigned short into half float in webgl2 we are losing precision. So we are saving memory at the cost of precision. So it is a trade off. Any thoughts on this?

Reply:
Thanks for pointing this out. I dug deeper into the half float standard, and according to [Wikipedia](https://en.wikipedia.org/wiki/Half-precision_floating-point_format) section "Precision limitations on integer values" :

- Integers between 0 and 2048 can be exactly represented (and also between −2048 and 0)
- Integers between 2048 and 4096 round to a multiple of 2 (even number)
- Integers between 4096 and 8192 round to a multiple of 4
- Integers between 8192 and 16384 round to a multiple of 8
- Integers between 16384 and 32768 round to a multiple of 16
- Integers between 32768 and 65519 round to a multiple of 32

Since at the time we convert the buffer, we should know the `min` and `max`, one approach can be to convert by default if the `min` and `max` are in the range that can be exactly represented (-2048 to 2048), and be Optional (by flag, default false) if otherwise?

Reply:
Actually there is another extension [EXT_texture_norm16](https://developer.mozilla.org/en-US/docs/Web/API/EXT_texture_norm16), but it is totally unclear to me if it makes the texture filterable. Any comment on this extension to enable non-Float32 3D textures that are filterable? The detail of this extension is [here](https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_texture_norm16.txt), check Table 8.13

Reply:
Per that table it would be filterable (TF == Texture Filterable) so it would work. However I'm not sure what systems have that extension. I checked on Chrome/Firefox on windows and neither reported it

I like the idea of both checking the data range and having a setting. The setting could be on the volume mapper and named preferSizeOverAccuracy or something like that.

Reply:
Sounds good, I'll add the `preferSizeOverAccuracy` for this PR.

Regarding the `EXT_texture_norm16`, I have it on my Chrome on Mac. I think it is GPU dependant, I checked mine on [WebGL report website](https://webglreport.com/?v=2)
Currently I'm working on a [demo](https://jsfiddle.net/sedghi/hxLegk5y/45/) to utilize `EXT_texture_norm16`, but I haven't had any luck so far, which is why I moved to half_float. However, I've begun a conversation with one of the developers of this extension, and I'll keep you updated, and hopefully create another PR later when the problem is resolved with `EXT_texture_norm16`.

Reply:
I believe the current code will cause all 2D textures created on WebGL2 to use half float event if their range exceeds -2048 to 2048. So I think we need to pass in an argument about using half float for short/ushort into getDefaultDataType maybe? Even if we just default to always false for 2D textures.

Reply:
@martinken Thanks for catching this, I fixed the logic. Do you have any comment for the `additiveBlend` in halfFloat which is not rendering the same thing as float?

Reply:
Since is no native support for 16 bit texture in webGL as far as I know, the options are 

1) use float32 (memory inefficient)

2) use norm16Texture (not widely available)

- https://github.com/Kitware/vtk-js/pull/2058

3) use half float (at the cost of inaccuracies in rendering, but almost widely available)

In WebGPU as I understand 16 bit texture is natively supported so vtk.js should not do that in webGPU. Otherwise, I can't think of any other solution.

Reply:
I guess you can put it under a flag to do half float when memory is the end goal, but i thought I added that here 

fix(imageMapper): not exact half float should be under a flag #2786

Reply:
Int16 and Uint16 textures are supported by WebGL so I didn't understand why all this was needed
But I found out that even if they are supported but they are not filterable, which means that we can only use nearest neighbor interpolation

Do you think that using software interpolation would be a good idea?
It would cost 8 unfiltered texture sampling and 7 call to `mix` but we don't have to do expensive operations to convert the volume, or loose memory

Reply:
They are not filterable correct. I mean I love to see software filtering on 16bits, but do you think doing 8 unfiltered texture sampling and 7 call to mix would be cheap? It might be, but if the fps is like 15 after I think we should let the user decide with some flags. But I love your idea

Reply:
I made [this small shader](https://www.shadertoy.com/view/X3SXRw) in ShaderToy to benchmark trilinear sampling in the fragment shader
It looks like trilinear interpolation in software is around 8.5 times slower than in hardware, but I don't know how many samples are made per pixel in the volume mapper
I suppose that it depends on the options, for example if the shadows are enabled